File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], t: int) -> int:
        n = len(nums)
        s = 0 
        e = n-1 
        mid = s + (e - s)//2 

        while s<= e:
            mid = s + (e - s)//2 
            if(nums[mid] == t):
                return mid

            elif(nums[mid] >= nums[s]):
                if(nums[s] <= t and nums[mid] > t):
                    e = mid - 1
                else :
                    s = mid + 1
            else:
                if(nums[mid] < t and nums[e] >= t):
                    s = mid + 1
                else:
                    e = mid - 1

        return -1

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int t) {
        int s = 0;
        int e = nums.size() - 1;
        int mid = s + (e - s)/2;
        while(s<=e){
            mid = s + (e - s)/2;

            if(nums[mid] == t)
                return mid;
            
            else if(nums[s]<= nums[mid]){
                if(nums[s]<=t && nums[mid]>t){
                    e = mid -1;
                }
                else{
                    s = mid +1;
                }
            }
            else{
                if(nums[mid]<t && t<=nums[e]){
                    s = mid +1;
                }
                else{
                    e = mid -1;
                }
            }
        }
        return -1;
    }
};